llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,435 @@
+    AU.addRequired<SlotIndexes>();
+    AU.addPreserved<SlotIndexes>();
+    AU.addRequired<LiveDebugVariables>();
+    AU.addPreserved<LiveDebugVariables>();
+    AU.addRequired<LiveStacks>();
+    AU.addPreserved<LiveStacks>();
+    AU.addRequired<LiveIntervals>();
+    AU.addPreserved<LiveIntervals>();
+    AU.addRequired<LiveRegMatrix>();
+    AU.addPreserved<LiveRegMatrix>();
+    AU.addRequired<MachineLoopInfo>();
+    AU.addPreserved<MachineLoopInfo>();
+    MachineFunctionPass::getAnalysisUsage(AU);
+  }
+
+  bool runOnMachineFunction(MachineFunction &Fn) override;
+
+private:
+  MachineFunction *MF = nullptr;
+  const MachineRegisterInfo *MRI = nullptr;
+  const AIEBaseRegisterInfo *TRI = nullptr;
+  VirtRegMap *VRM = nullptr;
+  LiveRegMatrix *LRM = nullptr;
+  LiveIntervals *LIS = nullptr;
+
+  bool
+  renameMBBPhysRegs(const SmallVector<const MachineBasicBlock *, 4> &LoopMBBs);
+  SmallSet<MCPhysReg, 16>
+  getMBBForbiddenReplacements(const MachineBasicBlock *MBB);
+
+  bool replaceReg(const Register Reg,
+                  SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs);
+
+  /// Find Free register of the same register Class type and
+  // exclude the forbidden Physical register from the result
+  SmallVector<MCPhysReg, 16>
+  getPriorityReplacementRegs(const Register Reg,
+                             const SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs);
+
+  bool isWAWCandidate(const Register &Reg, const MachineInstr &MI,
+                      const SmallVector<MCPhysReg, 16> &UsedPhyRegs,
+                      const SmallVector<MCPhysReg, 16> &UsedVirtRegs);
+
+  bool isIdentityCopy(const MachineInstr &MI);
+
+  // Check if the UnDef is the last define for a Register, or if there is
+  // another Def/Undef that will later complete the Definition of the
+  // Register. If so, it is too early to determine, if the Register is a
+  // candidate for a WAW replacement, or not. This is especially interesting,
+  // if Subregisters are used to hold a constant and this shared
+  // subregister-constant is equal for multiple Registers. A renamig here
+  // would destroy the shared usage of the sub-register and force the
+  // insertion of redundant Instructions.
+  bool isUndefReserved(
+      const MachineOperand &Def, const MachineInstr &MI,
+      const MachineBasicBlock &MBB,
+      std::map<unsigned, SmallVector<const MachineInstr *, 4>> &UndefPartReuse);
+
+  bool isLastDef(const MachineOperand &Def, const MachineInstr &DefInstr,
+                 const MachineBasicBlock &MBB);
+
+  bool reuseSubRegs(SmallVector<const MachineInstr *, 4> &OrigInstr,
+                    SmallVector<const MachineInstr *, 4> &OtherInstr);
+
+  bool containEqualSubRegDefs(const MachineInstr *InstrA,
+                              const MachineInstr *InstrB);
+
+  void addAliasRegs(SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs,
+                    const MCPhysReg &PhysReg);
+
+  Register getPhysReg(const MachineOperand &MO) const;
```

Comment:
Nit: I think `getAssignedPhysReg` would better describe what this function is doing.

Reply:
fixed